src/Serializer/IdentityMap.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace eLife\ApiSdk\Serializer;
+
+use GuzzleHttp\Promise\PromiseInterface;
+use function GuzzleHttp\Promise\all;
+
+/**
+ * As in http://www.martinfowler.com/eaaCatalog/identityMap.html.
+ */
+final class IdentityMap
+{
+    private $contents = [];
+
+    public function reset($id) : self
+    {
+        $this->contents[$id] = null;
+
+        return $this;
+    }
+
+    public function has($id) : bool
+    {
+        return array_key_exists($id, $this->contents);
+    }
+
+    /**
+     * @return PromiseInterface|null
+     */
+    public function get($id)
+    {
+        return $this->contents[$id];
+    }
+
+    public function fillMissingWith(callable $load) : self
+    {
+        foreach ($this->contents as $id => $promise) {
+            if (null === $promise) {
+                $this->contents[$id] = $load($id);
+            }
+        }
+
+        return $this;
+    }
+
+    public function waitForAll()
```

Comment:
I'd like to limit what can be done with this object rather than adding operations, should be an internal of Normalizers